RPM sensing from GPIO (PWM pin) pulses

MaEtUgR opened the pull request:
### Solved Problem
When wokring on helicopter rpm control I found that RPM sensing is "easily" possible but requires work. @bkueng provided the nice configurable capture part for that effort and I used the output directly in the heli-specific controller. But now I think we should enable generic RPM capturing for use in control, log, UI reporting.

### Solution
~As a draft, I added the hardcoded logic we used to process the RPM.
Note: This doesn't work well out of the box because it cannot detect a timeout if the interrupt is never called anymore.~
The solution in the meantime has:
- timeout
- outlier rejection
- pule per revolution configuration

Thanks for all the help from @oravla5, @sfuhrer and @bresch behind the scenes.

### Changelog Entry
```
Feature: Easy to use RPM sensing from GPIO pulses
```

### Test coverage
This logic was used on a heli to control RPM before porting it to latest main and moving the processing to the rpm capture driver. EDIT: and changing it completely again.

~We need to follow up with further testing, settling the interface, implementing the timeout... FYI @sfuhrer~
We did testing on an RC helicopter with high pulse per revolution count + noisy signal and a combustion engine with outliers.

sfuhrer commented:
We need to get better with rejecting outliers:
![image](https://github.com/user-attachments/assets/232dc14d-60ed-42fc-9539-defa6906994e)

oravla5 commented:
> We need to get better with rejecting outliers: ![image](https://private-user-images.githubusercontent.com/26798987/393325968-232dc14d-60ed-42fc-9539-defa6906994e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM4MjY2MTUsIm5iZiI6MTczMzgyNjMxNSwicGF0aCI6Ii8yNjc5ODk4Ny8zOTMzMjU5NjgtMjMyZGMxNGQtNjBlZC00MmZjLTk1MzktZGVmYTY5MDY5OTRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjEwVDEwMjUxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkOGIzYzk4MzViM2NkNzQxM2NhYjhkY2NiZjgxODU1NGViMTZmYTBhMTcwZTUyNmRiNzI0NTgzMThjOGQ2MzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.99i-6byRDfrkp2Zv5YFITmen5ueMdbPOM1Pduon8xuE)

@sfuhrer 
Do you by any chance still have this ulog?

MaEtUgR commented:
I rebased on main, fixed the screw up with the unwanted, unused `RpmControl.hpp` file I added and was changed for quick tests instead of the real logic. And added all our latest changes we came up with @oravla5.

> We did testing on an RC helicopter with high pulse per revolution count + noisy signal and a combustion engine with outliers.

Here's the plot from the small RC heli spinning the motor without blades open loop:
![grafik](https://github.com/user-attachments/assets/f23899d8-d2f0-4627-9b98-15c563f616fd)

oravla5 commented on src/drivers/rpm_capture/rpm_capture_params.c:
```diff
@@ -0,0 +1,55 @@
+/**
+ * RPM Capture Enable
+ *
+ * Enables the RPM capture module on FMU channel 5.
```
Before merging, shall we check if we could use other channels different from 5? If so, we should update the comment accordingly.

sfuhrer replied:
Yes definitely. I always used 5 so far because of that comment...

MaEtUgR replied:
I quickly asked the ST experts and the interrupt should be independent of timers and configuring the pin for input interrupt should be possible on most pins but probably depends on the board. So the assumption is it should work also on the other FMU PWM pins.

oravla5 replied:
Confirmed :ok_hand:

bkueng replied:
Channel 5 has history reasons. This is now more flexible as any pin can be configured, so the comment does not apply anymore.

oravla5 reviewed (commented):
LGTM once we check if we can do RPM sensing in other channels that are not 5.

oravla5 commented:
I confrimed that RPM capture works on other AUX GPIO pins. It's not available in the MAIN pins.

I also played around with the parameters of the alpha and medium filters:
- Increasing the median filter window makes the rpm estimate significantly smoother for the small heli noisy raw RPM, but also slower. A value of 7 seems to be a good compromise.
- Increasing the alpha filter factor does not have a significant effect.

Find attached the logs for:
- Alpha filter 0.5, median filter window 5
- Alpha filter 0.5, median filter window 11
- Alpha filter 0.5, median filter window 7
- Alpha filter 0.7, median filter window 7

![alpha0.5_median5](https://github.com/user-attachments/assets/46b3e68e-a92f-4bde-bfe5-825b19832cc7)
![alpha0.5_median11](https://github.com/user-attachments/assets/f366bb71-77c3-4c71-9409-92ae304f0422)
![alpha0.5_median7](https://github.com/user-attachments/assets/6afc9eb0-12fe-42f8-bda5-36b9d2eb325e)
![alpha0.7_median7](https://github.com/user-attachments/assets/ebf1c7ea-8f32-4b69-b4d2-3a66ae030b73)

[rpm_caputure_filter_tests.zip](https://github.com/user-attachments/files/18097699/rpm_caputure_filter_tests.zip)

bresch commented on src/drivers/rpm_capture/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+############################################################################
+#
+#   Copyright (c) 2021 PX4 Development Team. All rights reserved.
```
```suggestion
#   Copyright (c) 2024 PX4 Development Team. All rights reserved.
```

MaEtUgR replied:
It was in all the new files 😄

MaEtUgR replied:
a5f81a19f3f33c50aea23722798cb1ba76cf8a12

bresch commented on msg/Rpm.msg:
```diff
@@ -1,4 +1,5 @@
-float32 indicated_frequency_rpm       # indicated rotor Frequency in Revolution per minute
-float32 estimated_accurancy_rpm       # estimated accuracy in Revolution per minute
+float32 rpm_estimate # filtered revolutions per minute
+float32 rpm_raw # measured rpm
```
```suggestion
float32 rpm_raw
```

MaEtUgR replied:
a5f81a19f3f33c50aea23722798cb1ba76cf8a12

bresch commented on msg/Rpm.msg:
```diff
@@ -1,4 +1,5 @@
-float32 estimated_accurancy_rpm       # estimated accuracy in Revolution per minute
+float32 rpm_estimate # filtered revolutions per minute
+float32 rpm_raw # measured rpm
+float32 estimated_accurancy_rpm # estimated accuracy
```
```suggestion
float32 estimated_accurancy_rpm
```

MaEtUgR replied:
a5f81a19f3f33c50aea23722798cb1ba76cf8a12

bresch commented on src/drivers/rpm_capture/RPMCapture.cpp:
```diff
@@ -0,0 +1,220 @@
+{
+	bool success = false;
+
+	for (unsigned i = 0; i < 16; ++i) {
```
Could we have a constexpr instead of a magic number?

MaEtUgR replied:
31269c77f2cf6bbb586f02a22f06cbb3281295db

bresch commented on src/drivers/rpm_capture/RPMCapture.cpp:
```diff
@@ -0,0 +1,220 @@
+		int32_t function;
+
+		if (function_handle != PARAM_INVALID && param_get(function_handle, &function) == 0) {
+			if (function == 2070) { // RPM_Input
```
why 2070?

MaEtUgR replied:
It's defined as a function in this pr:
https://github.com/PX4/PX4-Autopilot/blob/bdec521b0a629d47d053a31d303f41913e355dc8/src/lib/mixer_module/output_functions.yaml#L64
I'm not sure if that's defined outside of the metadata 👀 otherwise I'll extend the comment.

MaEtUgR replied:
a5f81a19f3f33c50aea23722798cb1ba76cf8a12

bresch commented on src/drivers/rpm_capture/RPMCapture.cpp:
```diff
@@ -0,0 +1,220 @@
+		return false;
+	}
+
+	int ret = io_timer_allocate_channel(_channel, IOTimerChanMode_Other); // TODO: add IOTimerChanMode_RPM
```
Is it too difficult to add it now?

MaEtUgR replied:
d5f09596283d0d4c7ef128b7655acc1ac65e9848

bresch commented on src/drivers/rpm_capture/RPMCapture.cpp:
```diff
@@ -0,0 +1,220 @@
+		// Don't update RPM filter with outliers
+		const float dt = math::min((now - _timestamp_last_update) * 1e-6f, 1.f);
+		_timestamp_last_update = now;
+		_rpm_filter.setParameters(dt, 0.5f);
```
add a constexpr for "RPM filter time constant" ?

bresch commented on src/drivers/rpm_capture/RPMCapture.hpp:
```diff
@@ -0,0 +1,96 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
```
```suggestion
 *   Copyright (c) 2024 PX4 Development Team. All rights reserved.
```

MaEtUgR replied:
a5f81a19f3f33c50aea23722798cb1ba76cf8a12

bresch reviewed (changes requested):
Just a bit of nitpicking, otherwise looks good

MaEtUgR commented:
@bresch I rebased on main, addressed all your comments in new commits (not one each but they're referenced)
and added 375cd553b9b2014b5796a325676ec12b742307c7 on top because I saw that was only necessary before https://github.com/PX4/PX4-Autopilot/commit/15fece7e141ee8c3a3357ceb1f0e7b55f0566558#diff-21b8564105b699e09127da22888e7b35b845e53be24c30a6f921ea1f9e82f821L84.

MaEtUgR commented:
`v6xrt_allyes` 👀 
![grafik](https://github.com/user-attachments/assets/4719250e-7e2e-402a-8bfb-fb997e9c689c)

🙈 
![grafik](https://github.com/user-attachments/assets/cf00d02f-d11c-4488-88b5-a6fea1b83f09)

MaEtUgR commented:
`v6xrt_allyes` fixed by https://github.com/PX4/PX4-Autopilot/commit/de04cb1ac4053e60ac96c459b991fc857abbe9fc#diff-5cc1f83d05caa7e38df255ec8fb58eeac44362c55bca89ffcfe55de562b46c82R43

`holybro_kakutef7` fixed by https://github.com/PX4/PX4-Autopilot/pull/24126 (needs to be reviewed)

hamishwillee commented:
@MaEtUgR Can we have docs/info for this?

Specifically
- What kind of RPM sensors are supported? Do we have any sensors we can list that we know to work
- Is a tachometer sensor like this supported: https://docs.px4.io/main/en/sensor/tachometers.html
- That doc ^^^ says **Currently PX4 just logs RPM data: it is not used for state estimation or control.**. What use cases are we currently covering for this now - looks like helicopter RPM control (whatever that is) and ICE engines. Anything else?
- Given it is supported, is there any special setup required?
  - That tachometer suggest I2C as per https://docs.px4.io/main/en/sensor/thunderfly_tachometer.html#hardware-setup - but looking at this it looks like you require AUX/PWM input - or is that an "As well" and as long as it gets into the `Rpm` message you don't care?
- What does this support in helicopters that is now undocumented. I'm thinking updates to [helicopter types](https://docs.px4.io/main/en/frames_helicopter/#helicopter-types) and [Config](https://docs.px4.io/main/en/config_heli/)

Anything else you think I am missing!
